Remove helper traits

nox opened the pull request:
Now that `JSRef<T>` is gone, there is no need to have helper traits.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7401)

<!-- Reviewable:end -->

nox committed:
Remove helper traits

Now that JSRef<T> is gone, there is no need to have helper traits.

On components/script/*.rs:

    # Remove imports.
    /^ *use dom::[a-z]+::\{.*Helpers/ {
        s/\{(Raw[^L]|[^L][^a])[a-zA-Z]+Helpers, /\{/
        s/, (Raw[^L]|[^L][^a])[a-zA-Z]+Helpers([,}])/\2/g
        s/\{([a-zA-Z]+)\}/\1/
        /\{\}/d
        s/::self;$/;/
    }
    /^ *use dom::[a-z]+::\{?(Raw[^L]|[^L][^a])[a-zA-Z]+Helpers\}?;$/d

On components/script/dom/*.rs:

    # Ignore layout things.
    /^(pub )?(impl|trait).*Layout.* \{/,/^}$/ { P; D; }

    # Delete helpers traits.
    /^(pub )?trait ([^L][^ ]|L[^a])[^ ]+Helpers(<'a>)? \{$/,/^\}$/D

    # Patch private helpers.
    /^impl.*Private.*Helpers/,/^\}$/ {
        s/^impl<'a> Private([^L][^ ]|L[^a])[^ ]+Helpers(<'a>)? for &'a ([^ ]+) \{$/impl \3 {/
        /^ *(unsafe )?fn .*\(self.*[<&]'a/ {
            s/&'a /\&/g
            s/<'a, /</g
        }
        /^ *(unsafe )?fn /s/\(self([,)])/\(\&self\1/
    }

    # Patch public helpers.
    /^impl.*Helpers/,/^\}$/ {
        s/^impl(<'a>)? ([^L][^ ]|L[^a])[^ ]+Helpers(<'a>)? for (&'a )?([^ ]+) \{$/impl \5 {/
        /^ *(unsafe )?fn .*\(self.*[<&]'a/ {
            s/&'a /\&/g
            s/<'a, /</g
        }
        /^ *(unsafe )?fn .*\(&?self[,)]/s/(unsafe )?fn/pub &/
        /^ *pub (unsafe )?fn /s/\(self([,)])/\(\&self\1/
    }

The few error cases were then fixed by hand.

nox committed:
Remove dead code that surfaced when moved outside helper trait

nox committed:
Remove SinkHelpers

Method get_or_create() is now directly on servohtmlparser::Sink.

nox committed:
Remove reference to helper traits from documentation

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

frewsxcv commented:
@nox FYI, in case you missed it, this fails tidy

Manishearth commented on components/script/dom/attr.rs:
```diff
@@ -264,17 +256,17 @@ impl<'a> AttrHelpers<'a> for &'a Attr {
     }
 
-    fn value(self) -> Ref<'a, AttrValue> {
+    pub fn value<'a>(&'a self) -> Ref<'a, AttrValue> {
```
No need for using 'a anywhere in this line (`fn foo(&self)->Ref<Bar>` will work) because elision.

ditto everywhere else in this PR

Manishearth commented on components/script/dom/servohtmlparser.rs:
```diff
@@ -42,6 +44,19 @@ pub struct Sink {
 }
 
+impl Sink {
+    pub fn get_or_create(&self, child: NodeOrText<JS<Node>>) -> Root<Node> {
```
Where did this come from?

frewsxcv replied:
https://github.com/servo/servo/pull/7401/files#diff-14d82a613adf6dbeec0ec4415a1d323aL52

Manishearth replied:
Gotcha

Manishearth commented:
LGTM, though I'd prefer if elidable things were elided (clears up the code, and makes it easier to sed). I think all lifetime params introduced in this PR can be elided.

I may do a more thorough Reviewable review once I have a better viewport, but it otherwise looks good.

Ms2ger commented:
I much prefer not eliding lifetimes.

nox commented:
I prefer eliding too, gratuitous tokens that can be inferred don't please me. Will try to sed my way.

nox committed:
Remove AttributeHandlers

On components/script/*.rs:

    # Remove imports.
    /^ *use dom::element::\{.*AttributeHandlers/ {
        s/\{AttributeHandlers, /\{/
        s/, AttributeHandlers//g
        s/\{([a-zA-Z]+)\}/\1/
        /\{\}/d
        s/::self;$/;/
    }
    /^ *use dom::element::\{?AttributeHandlers\}?;$/d

    # Remove AttributeHandlers.
    /^pub trait AttributeHandlers \{$/,/^\}$/D

    # Patch AttributeHandlers methods.
    /^impl<'a> AttributeHandlers for &'a Element \{/,/^\}$/ {
        s/^impl<'a> AttributeHandlers for &'a Element \{/impl Element {/
        /^ *fn /s/\(self([,)])/\(\&self\1/
	/^ *fn.*\(&self/s/fn/pub fn/
    }

The few error cases were then fixed by hand.

nox commented:
Now with elision, scripts in commit messages, and new commits to remove AttributeHandlers.

nox commented:
And finally, tidy not failing.

Manishearth commented:
r+, squash anything if you want to

Manishearth commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit c4f7b60 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth c4f7b60e785018da5adac455da46646ff28029ca -->

Manishearth commented:
@Ms2ger elision happens in obvious cases (e.g. `fn foo<'a>(&'a T) -> &'a U` and `fn bar(&'a T, &'b U, &'c V)`), where `&'a T` could also be a `T<'a>`, etc. In these cases it's mostly noise -- if you're returning a reference from a reference it's obvious that the lifetimes must be the same. It also makes the code harder to grep or sed, and is overall jarring to the eyes.

I really think we should only be eliding when we have to, because it's those cases where the lifetimes are doing something nontrivial and often need to be looked at. Having lifetime params everywhere just makes these cases less prominent.

As part of the clippy cleanup I will be moving the rest of Servo to elision (clippy detects areas where the lifetimes are unnecessary), so if we don't want to do that we should discuss it now.

Ms2ger commented:
@bors-servo r-

Ms2ger commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit c4f7b60 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth c4f7b60e785018da5adac455da46646ff28029ca -->

larsbergstrom commented:
@bors-servo retry

nox commented:
@bors-servo r-

nox commented:
@bors-servo p=1 r+

bors-servo commented:
:pushpin: Commit c4f7b60 has been approved by `nox`

<!-- @bors-servo r=nox c4f7b60e785018da5adac455da46646ff28029ca -->

nox commented:
@bors-servo r- r=Manishearth

bors-servo commented:
:pushpin: Commit c4f7b60 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth c4f7b60e785018da5adac455da46646ff28029ca -->

nox commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 2a028f6 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 2a028f66a2e7ab56094cf856ebdc20bc49ab7d4d -->

nox commented:
Rebased against last master to see if it makes it happy.

Manishearth commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 2a028f6 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 2a028f66a2e7ab56094cf856ebdc20bc49ab7d4d -->

bors-servo commented:
:hourglass: Testing commit 2a028f6 with merge 9094297...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2891), [gonk](http://build.servo.org/builders/gonk/builds/2050), [linux1](http://build.servo.org/builders/linux1/builds/2901), [linux2](http://build.servo.org/builders/linux2/builds/2924), [mac1](http://build.servo.org/builders/mac1/builds/3155), [mac2](http://build.servo.org/builders/mac2/builds/3145), [mac3](http://build.servo.org/builders/mac3/builds/1098)

Manishearth commented:
\o/

(also yay, you broke all the PRs)
